pkg/util/validation/validation.go

```diff
@@ -242,24 +242,33 @@ var httpHeaderNameRegexp = regexp.MustCompile("^" + httpHeaderNameFmt + "$")
 // definition of a valid header field name (a stricter subset than RFC7230).
 func IsHTTPHeaderName(value string) []string {
 	if !httpHeaderNameRegexp.MatchString(value) {
-		return []string{RegexError(httpHeaderNameFmt, "X-Header-Name")}
+		return []string{RegexError(httpHeaderNameFmt, true, "X-Header-Name")}
 	}
 	return nil
 }
 
-const configMapKeyFmt = "\\.?" + dns1123SubdomainFmt
+const configMapKeyFmt = `\A[.\-_a-zA-Z0-9]+[^.]?\z`
+const doubleDotFmt = `\.{2}?`
+const trailingDotFmt = `\.\z`
 
 var configMapKeyRegexp = regexp.MustCompile("^" + configMapKeyFmt + "$")
+var doubleDotRegexp = regexp.MustCompile(doubleDotFmt)
+var trailingDotRegexp = regexp.MustCompile(trailingDotFmt)
 
-// IsConfigMapKey tests for a string that conforms to the definition of a
-// subdomain in DNS (RFC 1123), except that a leading dot is allowed
+// IsConfigMapKey tests for a string that is a valid key for a ConfigMap or Secret
 func IsConfigMapKey(value string) []string {
 	var errs []string
 	if len(value) > DNS1123SubdomainMaxLength {
 		errs = append(errs, MaxLenError(DNS1123SubdomainMaxLength))
 	}
 	if !configMapKeyRegexp.MatchString(value) {
-		errs = append(errs, RegexError(configMapKeyFmt, "key.name"))
+		errs = append(errs, RegexError(configMapKeyFmt, true, "key.name", "KEY_NAME", "key-name"))
```

Comment:
The double-dot prefix test is because we reserve ..* in projected volumes
for use by the system.  Specifically, we need it so we can do atomic
updates.

On Fri, Jul 29, 2016 at 11:32 PM, Ed Robinson <notifications@github.com>
wrote:

> In pkg/util/validation/validation.go
> <https://github.com/kubernetes/kubernetes/pull/25458#discussion_r72882955>
> :
>
> >  func IsConfigMapKey(value string) []string {
> >  	var errs []string
> >  	if len(value) > DNS1123SubdomainMaxLength {
> >  		errs = append(errs, MaxLenError(DNS1123SubdomainMaxLength))
> >  	}
> >  	if !configMapKeyRegexp.MatchString(value) {
> > -		errs = append(errs, RegexError(configMapKeyFmt, "key.name"))
> > +		errs = append(errs, RegexError(configMapKeyFmt, true, "key.name", "KEY_NAME", "key-name"))
>
> Furthermore I am not convinced that a double dot prefix is harmful so long
> as there is at least one other char. I think we just need to test for ..
> itself
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25458/files/7c312193e2543f4bc71b91e93346c058b340ae22#r72882955>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVM0UhqmiXCFdDJ_wrgzkpBdmy4apks5qau_3gaJpZM4IbwmI>
> .
>